### PR TITLE
ipn/ipnlocal: shutdown sshServer on tailscale down

### PR DESCRIPTION
**Pull request by maisem**
Also lazify SSHServer initialization to allow restarting the server on a
subsequent `tailscale up`

Signed-off-by: Maisem Ali <maisem@tailscale.com>

### CODE REVIEW

File: ipn/ipnlocal/local.go [hunk 1]
```diff
@@ -1969,16 +1964,23 @@ func (b *LocalBackend) setPrefsLockedOnEntry(caller string, newp *ipn.Prefs) {
 		cc.Login(nil, controlclient.LoginDefault)
 	}
 
+	if oldp.RunSSH && oldp.WantRunning {
```
Comment (bradfitz) [inline ipn/ipnlocal/local.go:1967]: if you add a method like `ShouldSSHBeUp()` on the prefs, then this is just:

````go
    if oldp.ShouldSSHBeUp() && !newp.ShouldSSHBeUp() {
     
````

.... which reads a bit nicer

File: ipn/ipnlocal/local.go [hunk 2]
```diff
@@ -1969,16 +1964,23 @@ func (b *LocalBackend) setPrefsLockedOnEntry(caller string, newp *ipn.Prefs) {
...
+		if !newp.WantRunning || !newp.RunSSH {
+			b.mu.Lock()
```
Comment (bradfitz) [inline ipn/ipnlocal/local.go:1969]: why re-acquiring the lock? can this be done before the unlock on 1930?

File: ipn/ipnlocal/local.go [hunk 3]
```diff
@@ -3367,11 +3369,28 @@ func (b *LocalBackend) DoNoiseRequest(req *http.Request) (*http.Response, error)
...
-	if b.sshServer == nil {
-		return errors.New("no SSH server")
+func (b *LocalBackend) sshServerOrInit() (_ SSHServer, err error) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	if b.sshServer != nil {
+		return b.sshServer, nil
+	}
+	if newSSHServer == nil {
+		return nil, errors.New("no SSH server")
```
Comment (bradfitz) [inline ipn/ipnlocal/local.go:3379]: "no SSH server support" ?

**Review by bradfitz — APPROVED** [review]

File: ipn/ipnlocal/local.go (hunk 1 above)
Comment (maisem) [reply ipn/ipnlocal/local.go]: Done

File: ipn/ipnlocal/local.go (hunk 2 above)
Comment (maisem) [reply ipn/ipnlocal/local.go]: Good catch, done.

**Pull request merged**